server/server.ts

```diff
@@ -2,21 +2,199 @@ import 'dotenv/config';
+    const response = await db.query(sql, params);
+    // for sign-in on sign-up
+    const user = response.rows[0];
+    const { userId } = user;
+    const payload = { firstName, lastName, username, userId };
+    const token = jwt.sign(payload, hashKey);
+    console.log('payload', payload);
+
+    res.status(201).json({ user: payload, token });
+  } catch (err) {
+    next(err);
+  }
+});
+
+type Auth = {
+  username: string;
+  password: string;
+};
+
+app.post('/api/auth/sign-in', async (req, res, next) => {
+  try {
+    const { username, password } = req.body as Partial<Auth>;
+    if (!username || !password) {
+      throw new ClientError(401, 'invalid login');
+    }
+    const sql = `
+                select * from "Users"
+                where "username" = $1;
+                `;
+    const response = await db.query(sql, [username]);
+    const user = response.rows[0];
+    if (!user) throw new ClientError(401, 'invalid login');
+    if (!(await argon2.verify(user.hashedPassword, password))) {
+      throw new ClientError(401, 'invalid login');
+    } else {
+      delete user.hashedPassword;
+      const payload = user;
+      const token = jwt.sign(payload, hashKey);
+      res.json({ user: payload, token });
+    }
+  } catch (err) {
+    next(err);
+  }
 });
 
+type MovieRating = {
+  movieId: number;
+  userId: number;
+  title: string;
+  summary: string;
+  link: string;
+  rating: number
+}
+
+app.post('/api/auth/movie-rating/add', authMiddleware, async (req, res, next) => {
+  try {
+    const { title, summary, link, rating } = req.body;
+    validateBody(title, 'Title');
+    validateBody(summary, 'Summary');
+    validateBody(link, 'Link');
+    if (!rating && rating < 1 && rating > 5) {
+  throw new ClientError(400, 'Rating must be a number between 1 and 5')
+}
+    const sql = `
+                insert into "Movies" ("userId", "title", "summary", "link", "rating")
+                values ($1, $2, $3, $4, $5)
+                returning *;
+                `;
+    const params = [req.user?.userId, title, summary, link, rating];
+    const response = await db.query(sql, params);
+    const movieRating = response.rows[0] as MovieRating;
+    res.status(201).json(movieRating);
+  } catch (err) {
+    next(err);
+  }
+});
+
+
+function validateMovieId(id: number): void {
+  if (!Number.isInteger(id) || id < 1) {
+    throw new ClientError(400, 'MovieId must be a positive integer');
+  }
+}
+
+app.put(
+  '/api/auth/movie-rating/:movieId/edit',
+  authMiddleware,
+  async (req, res, next) => {
+    try {
+      const movieId = Number(req.params.movieId);
+      validateMovieId(movieId);
+      const { title, summary, link, rating } = req.body;
+      validateBody(title, 'title');
+      validateBody(summary, 'Summary');
+      validateBody(link, 'Link');
+      if (!rating && rating < 1 && rating > 5) {
+        throw new ClientError(400, 'Rating must be a number between 1 and 5')
+      }
```

Comment:
What if the user doesn't want to edit the title, summary, link, etc? Should they have to supply the data for each field each time?